pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go

```diff
@@ -0,0 +1,156 @@
+		},
+		{
+			description: "Fails registration with conflicting plugin name",
+			returnErr:   fmt.Errorf("conflicting plugin name"),
+		},
+		{
+			description: "Successfully register plugin during initial traverse after plugin watcher restarts",
+			returnErr:   nil,
+		},
+		{
+			description: "Fails registration with conflicting plugin name during initial traverse after plugin watcher restarts",
+			returnErr:   fmt.Errorf("conflicting plugin name"),
+		},
+	}
+
+	callbackCount := 0
+	w.AddHandler(PluginType, func(name string, versions []string, sockPath string) error {
+		require.True(t, callbackCount <= (len(testCases)-1))
+		glog.Infof("receives plugin watcher callback for test: %s\n", testCases[callbackCount].description)
+		require.Equal(t, PluginName, name, "Plugin name mismatched!!")
+		require.Equal(t, []string{"v1beta1", "v1beta2"}, versions, "Plugin version mismatched!!")
+		// Verifies the grpcServer is ready to serve services.
+		_, conn, err := dial(socketPath)
+		require.Nil(t, err)
+		defer conn.Close()
+
+		// The plugin handler should be able to use any listed service API version.
+		v1beta1Client := v1beta1.NewExampleClient(conn)
+		v1beta2Client := v1beta2.NewExampleClient(conn)
+
+		// Tests v1beta1 GetExampleInfo
+		_, err = v1beta1Client.GetExampleInfo(context.Background(), &v1beta1.ExampleRequest{})
+		require.Nil(t, err)
+
+		// Tests v1beta1 GetExampleInfo
+		_, err = v1beta2Client.GetExampleInfo(context.Background(), &v1beta2.ExampleRequest{})
+		ret := testCases[callbackCount].returnErr
+		callbackCount++
+		//require.True(t, true)
+		return ret
+	})
+	err = w.Start()
+	require.NoError(t, err)
+
+	p := NewTestExamplePlugin() //first cbk
+	go func() {
+		p.Serve(socketPath)
+	}()
+	require.True(t, waitForPluginRegistrationStatus(t, p.registrationStatus))
+
+	// Trying to start a plugin service at the same socket path should fail
+	// with "bind: address already in use"
+	err = p.Serve(socketPath)
+	require.NotNil(t, err)
+
+	// grpcServer.Stop() will remove the socket and starting plugin service
+	// at the same path again should succeeds and trigger another callback.
+	p.Stop()
+	p = NewTestExamplePlugin() // 2nd cbk
+	go func() {
+		p.Serve(socketPath)
+	}()
+	require.True(t, waitForPluginRegistrationStatus(t, p.registrationStatus))
+
+	// Starting another plugin with the same name got verification error.
+	p2 := NewTestExamplePlugin() //3rd cbk
+	socketPath2 := socketDir + "/plugin2.sock"
+	go func() {
+		p2.Serve(socketPath2)
+	}()
+	require.False(t, waitForPluginRegistrationStatus(t, p2.registrationStatus))
+
+	// Restarts plugin watcher should traverse the socket directory and issues a
+	// callback for every existing socket.
+	err = w.Stop()
+	require.NoError(t, err)
+	errCh := make(chan error)
+	go func() {
+		errCh <- w.Start()
+	}()
+
+	require.True(t, waitForPluginRegistrationStatus(t, p.registrationStatus))
+	require.False(t, waitForPluginRegistrationStatus(t, p2.registrationStatus))
+
+	select {
```

Comment:
Shouldn't this be tested immediatly after starting? Why is the start in it's own goroutine ?Isn't it starting it's own goroutine?

Reply:
`traversePluginDir()`  --> `registerPlugin()`  --> `client.NotifyRegistrationStatus()` , this flow is before starting goroutine in start. If we test it just after start, there will be a deadlock because there will be no reading channel for registration status and start will not complete.